[receiver/vcenter] Add VM cpu usage and utilization

padraicbc opened the pull request:
**Description:** <Describe what has changed.>
Records the VM cpu usage and utilization in [recordVMUsages](https://github.com/padraicbc/opentelemetry-collector-contrib/blob/vm_cpu/receiver/vcenterreceiver/metrics.go#L76). 

**Testing:** 
Generated using `make mdatagen`

**Documentation:** 
Generated   [here](https://github.com/padraicbc/opentelemetry-collector-contrib/blob/vm_cpu/receiver/vcenterreceiver/documentation.md#vcentervmcpuusage) by `make mdatagen`

padraicbc committed:
vm cpu usage/util

padraicbc commented:
From a chat in the [vmware/govmomi](https://github.com/vmware/govmomi) slack room, the usage seems striaghtforward - the utlilzation seems a little more complicated so not 100% on that calculation.
From my understanding, unless explicit limts are set, the VM has the full cpu hz available in the host.  I couldn't find any field in that VM config that gave that informaton, so go it from the host directly.

schmikei reviewed (commented):
Thanks for the contribution!

Overall the additions look fairly reasonable and it was good to see your slack convo on the VMware slack, we have an integration test that should be able to come up with a value for this using the govmomi simulator which may need to be captured in the [integration test](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/vcenterreceiver/integration_test.go) and [integration test metrics file](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/vcenterreceiver/testdata/metrics/integration-metrics.json), but I'd like to spend some time validating around with my environment to see if the calculated values line up.

padraicbc commented:
> Thanks for the contribution!
> 
> Overall the additions look fairly reasonable and it was good to see your slack convo on the VMware slack, we have an integration test that should be able to come up with a value for this using the govmomi simulator which may need to be captured in the [integration test](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/vcenterreceiver/integration_test.go) and [integration test metrics file](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/vcenterreceiver/testdata/metrics/integration-metrics.json), but I'd like to spend some time validating around with my environment to see if the calculated values line up.

No worries at all. The usage aligns with what I see when looking in the VSphere UI and the util also seems to be close, but this is the first time using any of this so highly likely something is wrong :)

djaglowski commented on receiver/vcenterreceiver/metadata.yaml:
```diff
@@ -382,3 +382,19 @@ metrics:
+    attributes: []
+  vcenter.vm.cpu.usage:
+    enabled: true
+    description: The amount of CPU in Hz used by the VM.
```
```suggestion
    description: The amount of CPU used by the VM.
```
Generally, we don't include the units in the description, since it is documented as a separate field. In this case they did not match - exactly why we try not to specify twice.

djaglowski commented on receiver/vcenterreceiver/metadata.yaml:
```diff
@@ -382,3 +382,19 @@ metrics:
+  vcenter.vm.cpu.usage:
+    enabled: true
+    description: The amount of CPU in Hz used by the VM.
+    unit: "MHz"
```
Are you certain this unit makes sense? This is a measure of frequency (`cycles/second`). Often CPU usage is reported in `cycles`.

schmikei replied:
https://vdc-repo.vmware.com/vmwb-repository/dcr-public/1ef6c336-7bef-477d-b9bb-caa1767d7e30/82521f49-9d9a-42b7-b19b-9e6cd9b30db1/vim.vm.Summary.QuickStats.html

The `overallCpuUsage` field of this object in the VMware ObjectManager is what is being reported, so I think its accurate for MHz being the unit just based off this but just figured I'd chime in!

padraicbc replied:
I actually took the blocks from the host cpu usage and utliization in metadata.yaml, just changing host to vm. I checked various docs and megahertz is used across all for VM cpu usage. The vshpere client graph also shows _CPU usage in megahertz during the  interval: 1,234_.

padraicbc committed:
Update receiver/vcenterreceiver/metadata.yaml

Remove unit from desc

Co-authored-by: Daniel Jaglowski <jaglows3@gmail.com>

schmikei commented:
To follow up on my investigation onto the calculation, it appears that it seems fairly accurate across 10s of my Virtual Machines with active load, usually within +-.0005 of what the PerformanceManager states.  Seems like a fairly accurate/useful calculation from my perspective given the data.

![image](https://user-images.githubusercontent.com/32067685/225658154-567b9897-5418-4dab-af47-eb89c706dfa2.png)

padraicbc committed:
generate after removing unit

padraicbc commented:
> To follow up on my investigation onto the calculation, it appears that it seems fairly accurate across 10s of my Virtual Machines with active load, usually within +-.0005 of what the PerformanceManager states. Seems like a fairly accurate/useful calculation from my perspective given the data.
> 
> ![image](https://user-images.githubusercontent.com/32067685/225658154-567b9897-5418-4dab-af47-eb89c706dfa2.png)

Great. The next step is adding some cpu usage/util data to `integration-metrics.json` ?

schmikei commented:
> > To follow up on my investigation onto the calculation, it appears that it seems fairly accurate across 10s of my Virtual Machines with active load, usually within +-.0005 of what the PerformanceManager states. Seems like a fairly accurate/useful calculation from my perspective given the data.
> > ![image](https://user-images.githubusercontent.com/32067685/225658154-567b9897-5418-4dab-af47-eb89c706dfa2.png)
> 
> Great. The next step is adding some cpu usage/util data to `integration-metrics.json` ?

Yep it'll be adjusting the test suite, @djaglowski are you able to run CI for this PR to highlight which tests would potentially need to be updated for this change?

padraicbc committed:
Just use runtime.maxCpuUsage

padraicbc commented:
Turns out all I need is `runtime.maxCpuUsage` as it covers both cases, not just a limit scenario. I ran across many vm and `runtime.maxCpuUsage ` is indeed identical to `ncpu*hs.Summary.Hardware.CpuMhz`

padraicbc committed:
rebase

schmikei commented:
Calc and value looks good. We still need to modify the integration test to capture these changes I believe. I've created a commit here to elaborate what I mean

https://github.com/observIQ/opentelemetry-collector-contrib/commit/30a905d74965b7baf0c857367f9c41b7d94ec071

@djaglowski I believe if you ran CI you'd see failures on the integration-tests job. The content of that commit should resolve these issues.

padraicbc commented:
> Calc and value looks good. We still need to modify the integration test to capture these changes I believe. I've created a commit here to elaborate what I mean
> 
> [observIQ@30a905d](https://github.com/observIQ/opentelemetry-collector-contrib/commit/30a905d74965b7baf0c857367f9c41b7d94ec071)
> 
> @djaglowski I believe if you ran CI you'd see failures on the integration-tests job. The content of that commit should resolve these issues.

I had added some entries in the `testdata/metrics/integration-metrics.json`  and ran the end to end test locally but saw fails relating to the cpu usage. 
I added your commit in there and still see the same errors, one example is in `FluentForward-SplunkHEC`  where I see `ExpectedMaxCPU` set. 
What are the expected values calculated from?

padraicbc committed:
intg test

djaglowski commented:
> What are the expected values calculated from?

Typically the values are ignored in integration tests because they are unpredictable.

padraicbc commented:
> > What are the expected values calculated from?
> 
> Typically the values are ignored in integration tests because they are unpredictable.

Ah OK. I thought that had to pass. I missed an import in the last. It should be good now.

schmikei reviewed (approved):
Looks good to me. We still need a changelog entry describing the addition. Here is a link on how to accomplish this, but overall this is a great add. Thanks for the contribution! 🎉 

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/CONTRIBUTING.md#changelog

padraicbc committed:
chloggen

djaglowski commented on receiver/vcenterreceiver/testdata/metrics/expected.json:
```diff
@@ -1,4000 +1,4029 @@
 {
-    "resourceMetrics": [
+  "resourceMetrics": [
```
It seems this PR is changing the indentation of this file, which is making it difficult to understand the diff. Do you mind seeing if you can commit this without the indentation change?

padraicbc replied:
Yes,  that would be a lot easier 🤦‍♂️

padraicbc committed:
reformat expected.json

padraicbc committed:
fix json

padraicbc committed:
fix again..

djaglowski commented on receiver/vcenterreceiver/testdata/metrics/integration-metrics.json:
```diff
@@ -1,1052 +1,3125 @@
 {
-    "resourceMetrics": [
```
Do you mind doing this one too?

padraicbc replied:
I copied this from [here](https://github.com/observIQ/opentelemetry-collector-contrib/blob/30a905d74965b7baf0c857367f9c41b7d94ec071/receiver/vcenterreceiver/testdata/metrics/integration-metrics.json). I am not extirely sure what what is required. Can I checkout the main branch file and just add a single entry for both, or is there some criteria around what gets added and where?

djaglowski replied:
You can try to run the integration test locally by temporarily removing [these lines](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/024a11cd59dc6bddea1da9e06be16a1b665e8993/receiver/vcenterreceiver/integration_test.go#L15-L16). If that works, they you can use the `WriteMetrics` function to regenerate the whole file.

Otherwise, yeah, you can just start with what's on `main` and manually insert the changes. Ultimately, the goal is that this file represents the data you actually get when the receiver runs against the test container. (ignoring the unpredictable parts)

djaglowski reviewed (commented):
Thanks @padraicbc, this is much easier to follow.

padraicbc committed:
genrated intg

padraicbc commented:
> @gsanchezgavier, good catch. I'm able to run it locally if I change the name of the test to `TestIntegrationESX`, but otherwise I see the same behavior.
> 
> `make -C receiver/vcenterreceiver do-integration-tests-with-cover`
> 
> @schmikei, this is failing for me locally. Do you mind looking into this?

I removed the build tags and ran the test locally. The tests were non-deterministic with intermittent failures due to an unexpected number of values returned. The last json did however pass multiple times.

schmikei commented:
> @schmikei, this is failing for me locally. Do you mind looking into this?

I just pulled down the fork and it seems like the name change from `TestEndtoEnd_ESX` to `TestIntegrationESX` does work in running the integration test. The current version of the code I pulled down still noticed some issues when I ran it locally with the current expected integration metrics though @padraicbc 

<details>

<summary> test output </summary> 

```
        	            	 -  resource "map[vcenter.cluster.name:DC0_C0 vcenter.host.name:DC0_C0_H0 vcenter.vm.id:6132d223-1566-5921-bc3b-df91ece09a4d vcenter.vm.name:DC0_C0_RP0_VM1]": scope "otelcol/vcenterreceiver": number of metrics doesn't match expected: 11, actual: 12
        	            	 -  resource "map[vcenter.cluster.name:DC0_C0 vcenter.host.name:DC0_C0_H0]": scope "otelcol/vcenterreceiver": number of metrics doesn't match expected: 8, actual: 9
        	            	 -  resource "map[vcenter.cluster.name:DC0_C0 vcenter.host.name:DC0_C0_H0 vcenter.vm.id:b4689bed-97f0-5bcd-8a4c-07477cc8f06f vcenter.vm.name:DC0_H0_VM0]": scope "otelcol/vcenterreceiver": number of metrics doesn't match expected: 11, actual: 12
        	            	 -  resource "map[vcenter.cluster.name:DC0_C0 vcenter.host.name:DC0_C0_H0 vcenter.vm.id:12f8928d-f144-5c57-89db-dd2d0902c9fa vcenter.vm.name:DC0_H0_VM1]": scope "otelcol/vcenterreceiver": number of metrics doesn't match expected: 11, actual: 12
        	            	 -  resource "map[vcenter.cluster.name:DC0_C0 vcenter.host.name:DC0_C0_H0 vcenter.vm.id:bfff331f-7f07-572d-951e-edd3701dc061 vcenter.vm.name:DC0_C0_RP0_VM0]": scope "otelcol/vcenterreceiver": number of metrics doesn't match expected: 11, actual: 12
        	            	 -  resource "map[vcenter.cluster.name:DC0_C0 vcenter.host.name:DC0_C0_H0]": scope "otelcol/vcenterreceiver": number of metrics doesn't match expected: 8, actual: 9
        	            	 -  resource "map[vcenter.cluster.name:DC0_C0 vcenter.host.name:DC0_C0_H0]": scope "otelcol/vcenterreceiver": number of metrics doesn't match expected: 8, actual: 9
        	            	 
 ```
</details>

I did some light digging and it seems that the [govmomi](https://github.com/vmware/govmomi/commits/main/simulator) simulator package that we are using here now emits more performance metrics but since the test name was not getting picked up nobody noticed on a dependabot update that this got changed. The metric most noticeable to me was `vcenter.host.disk.latency.max` is now emitted from the simulator, whereas it was not captured in an earlier version of govmomi.

I think we should rely on CIs runs of the test and not individual local tests. 

@djaglowski do you think I should make a separate high priority Issue/PR that fixes running of the integration tests for this component by doing the rename.

padraicbc committed:
revert to host summary

djaglowski commented on receiver/vcenterreceiver/metrics.go:
```diff
@@ -71,6 +72,31 @@ func (v *vcenterMetricScraper) recordVMUsages(
+	var cpuUtilization float64
+
+	ncpu := vm.Config.Hardware.NumCPU
+	if cpuUsage != 0 {
```
When `cpuUsage == 0`, I think we shouldn't record the metric, unless we think it really is 0.

padraicbc replied:
You mean move the `RecordVcenterVMCPUUsageDataPoint`  and ` RecordVcenterVMCPUUtilizationDataPoint` inside the `if != 0`?

djaglowski replied:
I think that's appropriate, if you agree these "0"'s are meaningful

padraicbc replied:
Should there be cpu usage per vm in the simulator? Running it locally only gives me 0 for usage.

padraicbc committed:
delete json

padraicbc committed:
Merge remote-tracking branch 'upstream/main' into vm_cpu

padraicbc commented:
> @padraicbc I appreciate your patience with the golden files. Unlucky timing here, but the entire system was switched over from json to yaml as of a couple days ago. However, you should be able to regenerate the files using the same `WriteMetrics` call as before.

No worries. I was adding the call in `TestIntegrationESX` to write the `integration-metrics` file as that seemed to be the only place it was mentioned, and had the required args defined.  The `expected.yaml` is in the `scraper_test.go` but all tests in there fail due to ` POST "/sdk": 404 Not Found`.

padraicbc commented:
> > @padraicbc I appreciate your patience with the golden files. Unlucky timing here, but the entire system was switched over from json to yaml as of a couple days ago. However, you should be able to regenerate the files using the same `WriteMetrics` call as before.
> 
> No worries. I was adding the call in `TestIntegrationESX` to write the `integration-metrics` file as that seemed to be the only place it was mentioned, and had the required args defined. The `expected.yaml` is in the `scraper_test.go` but all tests in there fail due to ` POST "/sdk": 404 Not Found`.

I am running into the same 404 issue, running on a live system.

padraicbc commented:
> > I am running into the same 404 issue, running on a live system. The call to `host.Properties` to get the hardware summary seems to be the cause.
> 
> One tool I'd like you to add to your arsenal is [mitmproxy](https://mitmproxy.org/) I think govmomi abstracts away alot of errors into thinking its a 404 when its actually not. For example when not using `https://` you get a 404 even though actually its a 301. i.e.
> 
> ```
> 2023-04-13T14:21:02.946-0400	error	vcenterreceiver@v0.75.0/scraper.go:61	unable to establish a connection to the vSphere SDK unable to connect to vSphere SDK on listed endpoint: GET "/sdk": 404 Not Found	{"kind": "receiver", "name": "vcenter", "data_type": "metrics"}
> ```
> 
> <img alt="image" width="1644" src="https://user-images.githubusercontent.com/32067685/231864928-7b218ac0-22e2-48e5-b988-10912dcc0ad7.png">
> 
> I also did some digging into why you were getting those test failures, and its because of an unrouted call in the mockserver we used. We used `mitmproxy` to generate the XML that is returned from the SOAP API and did pretty manual routing for the test harness.
> 
> I created #20894 to supercede this PR and hopefully get the feature in with the least amount of hassle with updating the test harness. I'm still thinking about ways to make the testing of this component easier to add new metrics to, but right now we have something that "works" albeit a tad hard to update jumping in fresh.
> 
> Hopefully you are okay with this @padraicbc, want to highlight how patient you've been but wanted to get this feature in with less reviewer churn.
> 
> @djaglowski let me know if you're okay with reviewing/merging #20894 as a superceding PR or if you'd prefer @padraicbc to port the changes to this PR.

No worries.  The live failure is using an image I built with significant changes for the Kafka exporter too as it doesn't seem to support passing a proxy into the sarama config. I need that,  as one environment is all routed through a proxy.  That may well be causing some issues. I'll hopefully confirm tomorrow. 

I'm quite happy either way,  keeping the change separate in #20894 and rebasing off that once merged,  or adding it here.

schmikei commented:
> I also have vm memory util working locally. Should I add it to this PR or create a separate one?

My recommendation is to add a [feature request](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/new?assignees=&labels=enhancement%2Cneeds+triage&template=feature_request.yaml) as an issue for memory usage/utilization. Then follow up with a separate PR to keep commit history clean!

padraicbc commented:
Fwiw. I ran mitmproxy on the live system to debug. There was a redirect from http to https that likely caused the strange 404 sdk issue. It's working fine after changing to `https://vcenterip`  and setting a `no_proxy` env for that ip.
